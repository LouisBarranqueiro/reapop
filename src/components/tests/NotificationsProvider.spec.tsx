import React, {useEffect, useReducer} from 'react'
import {render} from '@testing-library/react'

import {NotificationsProvider, useNotifications} from '../..'
import mockReducer from '../../reducers/notifications/reducer'

const mockDispatch = jest.fn()

jest.mock('../../reducers/notifications/reducer', () => jest.fn())
jest.mock('react', () => {
    const react = jest.requireActual('react')
    return {
        ...react,
        useReducer: jest.fn((_, initialState) => {
            return [initialState, mockDispatch]
        }),
    }
})

describe('<NotificationsProvider/>', () => {
    it('should provide reducer state and actions', (done) => {
        const TestComponent = () => {
            const {notifications, notify, dismissNotification, dismissNotifications} = useNotifications()

            useEffect(() => {
                expect(useReducer).toHaveBeenNthCalledWith(1, mockReducer(), [])
                expect(notifications).toEqual([])

                const notification = notify({id: '1'})
                expect(notification).toMatchSnapshot()
                expect(mockDispatch.mock.calls).toMatchSnapshot()

                jest.clearAllMocks()
                dismissNotification('1')
                expect(mockDispatch.mock.calls).toMatchSnapshot()

                jest.clearAllMocks()
                dismissNotifications()
                expect(mockDispatch.mock.calls).toMatchSnapshot()

                done()
            }, [])

            return null
        }

        render(
            <NotificationsProvider>
                <TestComponent />
            </NotificationsProvider>
        )
    })
})
