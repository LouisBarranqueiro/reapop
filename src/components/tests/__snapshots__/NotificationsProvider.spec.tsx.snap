// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NotificationsProvider/> should provide reducer state and actions 1`] = `
Object {
  "buttons": Array [],
  "id": "1",
  "status": "none",
}
`;

exports[`<NotificationsProvider/> should provide reducer state and actions 2`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "buttons": Array [],
        "id": "1",
        "status": "none",
      },
      "type": "reapop/upsertNotification",
    },
  ],
]
`;

exports[`<NotificationsProvider/> should provide reducer state and actions 3`] = `
Array [
  Array [
    Object {
      "payload": "1",
      "type": "reapop/dismissNotification",
    },
  ],
]
`;

exports[`<NotificationsProvider/> should provide reducer state and actions 4`] = `
Array [
  Array [
    Object {
      "type": "reapop/dismissNotifications",
    },
  ],
]
`;
